auth_token/src/lib.rs

```diff
@@ -0,0 +1,97 @@
+#![no_std]
+
+#[cfg(feature = "testutils")]
+extern crate std;
+
+mod test;
+
+use soroban_auth::{
+    check_auth, NonceAuth, {Identifier, Signature},
+};
+use soroban_sdk::{contractimpl, contracttype, symbol, BigInt, Env, IntoVal, Map};
+
+#[contracttype]
+pub enum DataStoreKey {
+    Owner,
+    Admins,
+    Nonce(Identifier),
+}
+
+fn read_nonce(e: &Env, id: Identifier) -> BigInt {
+    let key = DataStoreKey::Nonce(id);
+    e.contract_data()
+        .get(key)
+        .unwrap_or_else(|| Ok(BigInt::zero(e)))
+        .unwrap()
+}
+
+struct NonceSignature(Signature);
+
+impl NonceAuth for NonceSignature {
+    fn read_nonce(e: &Env, id: Identifier) -> BigInt {
+        read_nonce(e, id)
+    }
+
+    fn read_and_increment_nonce(&self, e: &Env, id: Identifier) -> BigInt {
+        let key = DataStoreKey::Nonce(id.clone());
+        let nonce = Self::read_nonce(e, id);
+        e.contract_data().set(key, &nonce + 1);
+        nonce
+    }
```

Comment:
Something I find interesting about this is that even though the sdk tries to own nonces to a degree, and prescribes how they're used, this contract is still intimately involved. The SDK doesn't really abstract or remove much from the developer, but it does enforce a nonce existing that has a specific shape and behavior. For me this reinforces the idea noted here: https://github.com/stellar/rs-soroban-sdk/issues/533#issuecomment-1235885483.